src/line_parser2.rs

```diff
@@ -0,0 +1,390 @@
+use crate::{BeatNumber, Line, LineNote, Result};
+use combine::{
+    choice, many, many1, optional,
+    parser::char::{digit, space, spaces},
+    sep_by, token, Parser,
+};
+
+#[derive(Clone, Copy, Debug)]
+enum Letter {
+    A,
+    B,
+    C,
+    D,
+    E,
+    F,
+    G,
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Modifier {
+    Flat,
+    Natural,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct Note {
+    letter: Letter,
+    modifier: Modifier,
+    octave: i8,
+    duration: u32,
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Value {
+    Note(Note),
+    Rest,
+}
+
+impl Note {
+    fn to_wmidi_note(self) -> wmidi::Note {
+        let letter_value = match self.letter {
+            Letter::C => 0,
+            Letter::D => 2,
+            Letter::E => 4,
+            Letter::F => 5,
+            Letter::G => 7,
+            Letter::A => 9,
+            Letter::B => 11,
+        };
+
+        let modifier_value = match self.modifier {
+            Modifier::Flat => -1,
+            Modifier::Natural => 0,
+        };
+
+        let octave_value = self.octave + 1;
+
+        let value = octave_value * 12 + letter_value + modifier_value;
+        wmidi::Note::from_u8_lossy(value as u8)
+    }
+}
+
+fn parse_line(line: &str) -> Result<Line> {
+    let letter_parser = choice((
+        token('A').map(|_| Letter::A),
+        token('B').map(|_| Letter::B),
+        token('C').map(|_| Letter::C),
+        token('D').map(|_| Letter::D),
+        token('E').map(|_| Letter::E),
+        token('F').map(|_| Letter::F),
+        token('G').map(|_| Letter::G),
+    ));
+
+    let flat_parser = token('b').map(|_| Modifier::Flat);
+
+    let modifier_parser =
+        optional(flat_parser).map(|modifier| modifier.unwrap_or(Modifier::Natural));
+
+    let octave_parser = digit().map(|c| c.to_string().parse::<i8>().unwrap());
+
+    let pitch_parser = (letter_parser, modifier_parser, octave_parser, spaces())
+        .map(|(letter, modifier, octave, _)| (letter, modifier, octave));
+
+    let dot_parser = (token('.'), spaces()).map(|_| ());
+
+    let duration_parser = many(dot_parser).map(|dots: Vec<_>| (dots.len() + 1) as u32);
+
+    let note_parser =
+        (pitch_parser, duration_parser).map(|((letter, modifier, octave), duration)| {
+            Value::Note(Note {
+                letter,
+                modifier,
+                octave,
+                duration,
+            })
+        });
+
+    let rest_parser = (token('-'), spaces()).map(|_| Value::Rest);
+
+    let value_parser = choice((note_parser, rest_parser));
+
+    let mut line_parser = many1(value_parser).map(|notes: Vec<_>| to_line(&notes));
```

Comment:
What is `many1()` (vs `many()`)?

Also here, it seems like `to_line()` could consume `notes` (vs take a reference to it)?

Reply:
`many1` requires at least 1, `many` is 0 or more.